src/libtiled/tilelayer.h

```diff
@@ -331,19 +386,16 @@ class TILEDSHARED_EXPORT TileLayer : public Layer
 
     TileLayer *clone() const override;
 
-    // Enable easy iteration over cells with range-based for
-    QVector<Cell>::iterator begin() { return mGrid.begin(); }
-    QVector<Cell>::iterator end() { return mGrid.end(); }
-    QVector<Cell>::const_iterator begin() const { return mGrid.begin(); }
-    QVector<Cell>::const_iterator end() const { return mGrid.end(); }
+    void copyGrid(const QVector<Cell> &newGrid);
 
 protected:
     TileLayer *initializeClone(TileLayer *clone) const;
 
 private:
     int mWidth;
     int mHeight;
-    QVector<Cell> mGrid;
+    Cell mEmptyCell;
+    QMap< QPair<int, int>, Block* > mMap;
```

Comment:
Based on the `qHash` implementation for `QPair`, we can derive the following implementation for `QPoint`:
```c++
inline uint qHash(const QPoint &key, uint seed = 0) Q_DECL_NOTHROW
{
    uint h1 = qHash(key.x(), seed);
    uint h2 = qHash(key.y(), seed);
    return ((h1 << 16) | (h1 >> 16)) ^ h2 ^ seed;
}
```